### PR TITLE
Add support and example for picking other Three.js objects

### PR DESCRIPTION
**Pull request by dragly**
This changes our picking implementation to return an object similar to what Raycaster returns in Three.js. In addition, it changes the API to take a list of CadNode objects for picking that will be rendered individually. This leaves the distance comparison up to the user, but at least allows us to render to depth and tree index buffers without interference from other objects.

### CODE REVIEW

File: viewer/src/views/threejs/cad/picking.ts [hunk 1]
```diff
@@ -5,18 +5,135 @@
...
+    .fromArray(rgbaBuffer)
+    .multiplyScalar(1 / 255)
+    .dot(unpackFactors);
+}
+
+function perspectiveDepthToViewZ(invClipZ: number, near: number, far: number) {
+  return (near * far) / ((far - near) * invClipZ - far);
+}
+
+export function pickDepth(input: TreeIndexPickingInput): number {
```
Comment (larsmoa) [inline viewer/src/views/threejs/cad/picking.ts]: Not used outside picking.ts, don't export.

**Comment by larsmoa** [comment]
Is it plausible to unit test this?

File: viewer/src/views/threejs/cad/picking.ts [hunk 2]
```diff
@@ -5,18 +5,135 @@
...
+  for (const cadNode of cadNodes) {
+    const result = pickCadNode(cadNode, input);
+    if (result) {
+      results.push(result);
+    }
+  }
+  return results;
+}
+
+export function pickCadNode(cadNode: CadNode, input: IntersectCadNodesInput): TreeIndexPickingResult | undefined {
```
[the remark this answers is not in the document]
Comment (dragly) [reply viewer/src/views/threejs/cad/picking.ts]: This is to align with the API of RayCaster which has intersectObject and intersectObjects. I renamed it to intersectCadNode, though.

File: viewer/src/views/threejs/cad/picking.ts [hunk 3]
```diff
@@ -5,18 +5,135 @@
...
+export interface IntersectCadNodesInput {
+  coords: {
+    x: number;
+    y: number;
+  };
+  camera: THREE.PerspectiveCamera;
+  renderer: THREE.WebGLRenderer;
+}
+
+export interface IntersectCadNodesResult {
```
[the remark this answers is not in the document]
Comment (dragly) [reply viewer/src/views/threejs/cad/picking.ts]: Done

File: viewer/src/views/threejs/cad/picking.ts [hunk 4]
```diff
@@ -5,18 +5,135 @@
 import { CadNode } from './CadNode';
 import { pickPixelColor, PickingInput } from '../pickPixelColor';
 import { RenderMode } from '../materials';
+import * as THREE from 'three';
 
-interface TreeIndexPickingInput extends PickingInput {
+export interface TreeIndexPickingInput extends PickingInput {
```
[the remark this answers is not in the document]
Comment (dragly) [reply viewer/src/views/threejs/cad/picking.ts]: Should it not be exported all the way so that users can type the input?

**Comment by dragly** [comment]
> Is it plausible to unit test this?

I think it requires visual tests. We can only test the helper functions, but I think they are mostly an implementation detail and would rather test the entire picking pipeline.

**Pull request merged**